src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -1552,18 +1715,181 @@ CHIP_ERROR GroupDataProviderImpl::RemoveFabric(chip::FabricIndex fabric_index)
+    size_t count = 0;
+
+    for (size_t i = 0; i < mFabricTotal; i++, fabric.fabric_index = fabric.next)
+    {
+        if (CHIP_NO_ERROR != fabric.Load(mProvider.mStorage))
+        {
+            break;
+        }
+
+        // Iterate key sets
+        KeyMapData mapping(fabric.fabric_index, fabric.first_map);
+
+        // Look for the target group in the fabric's keyset-group pairs
+        for (uint16_t j = 0; j < fabric.map_count; ++j, mapping.id = mapping.next)
+        {
+            if (CHIP_NO_ERROR != mapping.Load(mProvider.mStorage))
+            {
+                break;
+            }
+
+            // Group found, get the keyset
+            KeySet keyset;
+            if (CHIP_NO_ERROR != mProvider.GetKeySet(fabric.fabric_index, mapping.keyset_id, keyset))
+            {
+                break;
+            }
+            ByteSpan epoch_key = keyset.GetCurrentKey();
+            if (epoch_key.empty())
+            {
+                continue;
+            }
+
+            uint8_t operational_key[Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
+            MutableByteSpan key_span(operational_key, sizeof(operational_key));
+
+            if (CHIP_NO_ERROR == Crypto::DeriveGroupOperationalKey(epoch_key, key_span))
+            {
+                uint16_t session_id = 0;
+                if (CHIP_NO_ERROR == Crypto::DeriveGroupSessionId(key_span, session_id) && session_id == mSessionId)
+                {
+                    count++;
+                }
+            }
+        }
+    }
+
+    return count;
+}
+
+bool GroupDataProviderImpl::GroupSessionIteratorImpl::Next(GroupSession & output)
+{
+    while (mFabricCount < mFabricTotal)
+    {
+        FabricData fabric(mFabric);
+        VerifyOrReturnError(CHIP_NO_ERROR == fabric.Load(mProvider.mStorage), false);
+
+        if (mMapCount >= fabric.map_count)
+        {
+            // No more keyset/group mappings on the current fabric, try next fabric
+            mFabric = fabric.next;
+            mFabricCount++;
+            mFirstMap = true;
+            mMapCount = 0;
+            continue;
+        }
+
+        if (mFirstMap)
+        {
+            mMapping  = fabric.first_map;
+            mFirstMap = false;
+        }
+
+        KeyMapData mapping(mFabric, mMapping);
+        VerifyOrReturnError(CHIP_NO_ERROR == mapping.Load(mProvider.mStorage), false);
+        mMapping = mapping.next;
+        mMapCount++;
+
+        // Group found, get the keyset
+        KeySet keyset;
+        VerifyOrReturnError(CHIP_NO_ERROR == mProvider.GetKeySet(fabric.fabric_index, mapping.keyset_id, keyset), false);
+
+        ByteSpan epoch_key = keyset.GetCurrentKey();
+        if (epoch_key.empty())
+        {
+            continue;
+        }
+
+        uint8_t operational_key[Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
+        MutableByteSpan key_span(operational_key, sizeof(operational_key));
+
+        if (CHIP_NO_ERROR == Crypto::DeriveGroupOperationalKey(epoch_key, key_span))
```

Comment:
We do not need to store the epoch key when we install a group key with Group Key Management cluster because there is no read operation. Therefore, we can save *all* of this derivation by doing it in advance when persisting a group key set being installed. Instead of storing the epoch key, immediately derive the group operational key, and store that. This will later save you all of this work.

Reply:
Great suggestion, now the epoch keys are not stored at all, instead, the operational group keys and hash are stored instead.